src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/nodes/process/WakeHighestPriorityNode.java

```diff
@@ -41,16 +43,18 @@ public void executeWake(final VirtualFrame frame) {
         // Return the highest priority process that is ready to run.
         // Note: It is a fatal VM error if there is no runnable process.
         final ArrayObject schedLists = pointersReadNode.executeArray(image.getScheduler(), PROCESS_SCHEDULER.PROCESS_LISTS);
-        long p = arraySizeNode.execute(schedLists) - 1;  // index of last indexable field
-        PointersObject processList;
-        do {
-            if (p < 0) {
-                errorProfile.enter();
-                throw SqueakException.create("scheduler could not find a runnable process");
+        for (long p = arraySizeNode.execute(schedLists) - 1; p >= 0; p--) {
+            final PointersObject processList = (PointersObject) arrayReadNode.execute(schedLists, p);
+            while (!processList.isEmptyList(pointersReadNode)) {
+                final PointersObject newProcess = processList.removeFirstLinkOfList(pointersReadNode, pointersWriteNode);
+                final Object newContext = pointersReadNode.execute(newProcess, PROCESS.SUSPENDED_CONTEXT);
+                if (newContext instanceof ContextObject) {
+                    contextNode.executeGet(frame).transferTo(pointersReadNode, pointersWriteNode, newProcess);
+                    throw SqueakException.create("Should not be reached");
+                }
             }
-            processList = (PointersObject) arrayReadNode.execute(schedLists, p--);
-        } while (processList.isEmptyList(pointersReadNode));
-        final PointersObject newProcess = processList.removeFirstLinkOfList(pointersReadNode, pointersWriteNode);
-        contextNode.executeGet(frame).transferTo(pointersReadNode, pointersWriteNode, newProcess);
+        }
+        errorProfile.enter();
+        throw SqueakException.create("scheduler could not find a runnable process");
```

Comment:
I cannot tell whether the logic has changed here at all, or if you only replaced the `do{} while()` here with a `for` loop. In case you're wondering, we used `do` to be closer to the implementation in Slang and [SqueakJS](https://github.com/bertfreudenberg/SqueakJS/blob/96edc18e3fcaf4e637f2978b8f0b6c3c608d961f/vm.js#L5796-L5807).

Reply:
You are right, I was mislead by some code in the image, ProcessorScheduler>>nextReadyProcess (used by the profiler), where if the first element happens to not have a suspended context, it is skipped and the iteration continues.
I should have checked Slang instead

Reply:
Actually, now that I did read what Slang does (in StackInterpreter>>wakeHighestPriority), the change was actually correct:
There is an explicit comment in the loop
			"Only answer processes with a runnable suspendedContext.
			  Discard those that aren't; the VM would crash otherwise."
...and then there is a check for a live context
			 (self isLiveContext: ctxt) ifTrue:
which was missing

Reply:
Ok, great! Thanks for double checking.